engine/shaderc/src/test/test_shaderc.cpp

```diff
@@ -0,0 +1,418 @@
+
+    dmShaderc::HShaderContext shader_ctx = dmShaderc::NewShaderContext(data, data_size);
+    const dmShaderc::ShaderReflection* reflection = dmShaderc::GetReflection(shader_ctx);
+
+    ASSERT_EQ(3, reflection->m_UniformBuffers.Size());
+
+    ASSERT_EQ(dmHashString64("ubo_global_one"), reflection->m_UniformBuffers[0].m_NameHash);
+    ASSERT_EQ(0,                                reflection->m_UniformBuffers[0].m_InstanceNameHash);
+
+    ASSERT_EQ(dmHashString64("ubo_global_two"), reflection->m_UniformBuffers[1].m_NameHash);
+    ASSERT_EQ(0,                                reflection->m_UniformBuffers[1].m_InstanceNameHash);
+
+    ASSERT_EQ(dmHashString64("ubo_inst"),       reflection->m_UniformBuffers[2].m_NameHash);
+    ASSERT_EQ(dmHashString64("inst_variable"),  reflection->m_UniformBuffers[2].m_InstanceNameHash);
+
+    dmShaderc::DeleteShaderContext(shader_ctx);
+}
+
+static const dmShaderc::ShaderResource* GetShaderResourceByNameHash(const dmArray<dmShaderc::ShaderResource>& resource_list, dmhash_t name_hash)
+{
+    for (uint32_t i = 0; i < resource_list.Size(); ++i)
+    {
+        if (resource_list[i].m_NameHash == name_hash)
+        {
+            return &resource_list[i];
+        }
+    }
+    return 0;
+}
+
+static inline const char* FindFirstOccurance(const char* src, const char* text)
+{
+    return strstr(src, text);
+}
+
+TEST(Shaderc, ModifyBindings)
+{
+    uint32_t data_size;
+    void* data = ReadFile("./build/src/test/data/bindings.spv", &data_size);
+    ASSERT_NE((void*) 0, data);
+
+    dmShaderc::HShaderContext shader_ctx = dmShaderc::NewShaderContext(data, data_size);
+    const dmShaderc::ShaderReflection* reflection = dmShaderc::GetReflection(shader_ctx);
+
+    const dmShaderc::ShaderResource* position  = GetShaderResourceByNameHash(reflection->m_Inputs, dmHashString64("position"));
+    const dmShaderc::ShaderResource* normal    = GetShaderResourceByNameHash(reflection->m_Inputs, dmHashString64("normal"));
+    const dmShaderc::ShaderResource* tex_coord = GetShaderResourceByNameHash(reflection->m_Inputs, dmHashString64("tex_coord"));
+
+    dmShaderc::HShaderCompiler compiler = dmShaderc::NewShaderCompiler(shader_ctx, dmShaderc::SHADER_LANGUAGE_GLSL);
+
+#if 0
+    dmShaderc::DebugPrintReflection(reflection);
+#endif
+
+    ASSERT_NE((void*) 0, position);
+    ASSERT_EQ(0, position->m_Location);
+
+    ASSERT_NE((void*) 0, normal);
+    ASSERT_EQ(1, normal->m_Location);
+
+    ASSERT_NE((void*) 0, tex_coord);
+    ASSERT_EQ(2, tex_coord->m_Location);
+
+    dmShaderc::SetResourceLocation(shader_ctx, compiler, dmHashString64("position"), 3);
+    dmShaderc::SetResourceLocation(shader_ctx, compiler, dmHashString64("normal"), 4);
+    dmShaderc::SetResourceLocation(shader_ctx, compiler, dmHashString64("tex_coord"), 5);
+
+    dmShaderc::SetResourceBinding(shader_ctx, compiler, dmHashString64("matrices"), 3);
+    dmShaderc::SetResourceBinding(shader_ctx, compiler, dmHashString64("extra"), 4);
```

Comment:
This will be useful for coming tasks. Vertex and fragment shaders are compiled to SPIRv separately, which means that locations and bindings for the same resources can have different values, which we can post-fix with these functions. We might need to take an extra compilation step to fixup the .spv file from the .fp and .vp shaders before crosscompiling, but it will fix all the issues with bindings we have.